electrumx/lib/coins.py

```diff
@@ -78,6 +78,13 @@ class Coin(object):
     PEER_DEFAULT_PORTS = {'t': '50001', 's': '50002'}
     PEERS = []
 
+    @classmethod
+    def txin_is_spendable(cls, tx_input):
+        """Some coins have implemented stealth transactions
+        In those cases, the script will be marked with OP_RETURN
+        and should be treated as not spendable."""
```

Comment:
The code already treats OP_RETURN as not spendable.  What makes you think otherwise?

Reply:
I feel like I have done a poor job at describing the initial issue. Apologies.

> The code already treats OP_RETURN as not spendable. What makes you think otherwise?


It does so for tx outputs (a check that is also called indirectly during the loop). For tx inputs it does not.

I changed the PR a bit to demonstrate the issue. Essentially, I removed the continue clause and added a log message including the script of the input. The diff compared to what is proposed here, is:

```diff
diff --git a/electrumx/server/block_processor.py b/electrumx/server/block_processor.py
index b895a9e..079c2ec 100644
--- a/electrumx/server/block_processor.py
+++ b/electrumx/server/block_processor.py
@@ -18,6 +18,7 @@ from functools import partial
 from electrumx.server.daemon import DaemonError
 from electrumx.lib.hash import hash_to_str, HASHX_LEN
 from electrumx.lib.util import chunks, formatted_time, class_logger
+from electrumx.lib.script import OpCodes
 import electrumx.server.db
 
 
@@ -535,8 +536,9 @@ class BlockProcessor(electrumx.server.db.DB):
             if not tx.is_coinbase:
                 for txin in tx.inputs:
                     # Verify that this is not a stealth input
-                    if self.coin.txin_is_spendable(txin) is False:
-                        continue
+                    if txin.script[0] == OpCodes.OP_RETURN:
+                        self.logger.warning("Found non spendable input: %r" % (txin.script,))
+                        #continue
                     cache_value = spend_utxo(txin.prev_hash, txin.prev_idx)
                     undo_info_append(cache_value)
                     append_hashX(cache_value[:-12])
@@ -617,8 +619,9 @@ class BlockProcessor(electrumx.server.db.DB):
             if not tx.is_coinbase:
                 for txin in reversed(tx.inputs):
                     # Verify that this is not a stealth input
-                    if self.coin.txin_is_spendable(txin) is False:
-                        continue
+                    if txin.script[0] == OpCodes.OP_RETURN:
+                        self.logger.warning("Found non spendable input: %r" % (txin.script,))
+                        #continue
                     n -= undo_entry_len
                     undo_item = undo_info[n:n + undo_entry_len]
                     put_utxo(txin.prev_hash + s_pack('<H', txin.prev_idx),
```

I removed the contents of the DB folder and ran the daemon again, forcing a sync. The result is as follows:

```log
$ electrumx_server 
WARNING:Env:lowered maximum sessions from 1,000 to 674 because your open file limit is 1,024
INFO:Controller:software version: ElectrumX 1.5
INFO:Controller:aiorpcX version: 0.5.6
INFO:Controller:supported protocol versions: 1.1-1.3
INFO:Controller:event loop policy: None
INFO:LegacyRPCDaemon:daemon #1 at 127.0.0.1:8800/ (current)
INFO:BlockProcessor:switching current directory to /home/electrum/electrum_db/
INFO:BlockProcessor:using leveldb for DB backend
INFO:BlockProcessor:created new database
INFO:BlockProcessor:creating metadata directory
INFO:History:flush count: 0
INFO:History:history DB version: 0
INFO:BlockProcessor:DB version: 6
INFO:BlockProcessor:coin: TokenPay
INFO:BlockProcessor:network: mainnet
INFO:BlockProcessor:height: -1
INFO:BlockProcessor:tip: 0000000000000000000000000000000000000000000000000000000000000000
INFO:BlockProcessor:tx count: 0
INFO:BlockProcessor:sync time so far: 00s
INFO:BlockProcessor:reorg limit is 500 blocks
INFO:BlockProcessor:flushing DB cache at 1,200 MB
INFO:Controller:RPC server listening on localhost:8000
INFO:Prefetcher:catching up to daemon height 241,815 (241,816 blocks behind)
INFO:Prefetcher:verified genesis block with hash 000008b71ab32e585a23f0de642dc113740144e94c0ece047751e9781f953ae9
INFO:BlockProcessor:our height: 9 daemon: 241,815 UTXOs 0MB hist 0MB
INFO:BlockProcessor:our height: 5,009 daemon: 241,815 UTXOs 2MB hist 2MB
INFO:BlockProcessor:our height: 10,009 daemon: 241,815 UTXOs 4MB hist 2MB
INFO:BlockProcessor:our height: 15,009 daemon: 241,815 UTXOs 6MB hist 3MB
INFO:BlockProcessor:our height: 20,009 daemon: 241,815 UTXOs 8MB hist 4MB
INFO:BlockProcessor:our height: 30,009 daemon: 241,815 UTXOs 13MB hist 5MB
INFO:BlockProcessor:our height: 37,509 daemon: 241,815 UTXOs 17MB hist 6MB
INFO:BlockProcessor:our height: 47,009 daemon: 241,815 UTXOs 22MB hist 9MB
INFO:BlockProcessor:our height: 54,509 daemon: 241,815 UTXOs 25MB hist 10MB
WARNING:BlockProcessor:Found non spendable input: b'j\xb9\xb0\x9bp\x0b\xea:N\xf3\xf7\xfe\xb9}\xcc!\x8d\xbcB\xcb\xb8\xc7w\x02\x13\xcc\xfd\x1f\xd5\xb1\x90\xc8\x8cug{\xf2\xb33\x0bmr\x97\xb8\xb3J\xd8\xb5\xb3d\x10\xad\x0e\xd3\xee2\x06\xdb\x8d8Z\x13\xf8>\xd8\xe0\x99!s^oY\x8a\xc88.c\x04\xbc\xec\x01 \xc7\x90\xb3\'\xd8\xc3\x94\x8e\x04b \xbbi\x1b3M\xban\xba?\xc0\x1e\xdb\xecK\x95\xd8}\xc52\xb0\x0b\xea\xbb\xf4$\x8c\xcc/O\x88\x84\xf1w\x85B\xb4\xe3G\xe4Z\xae\x9f\x19\xf2\x16#\xac1\xec\xbc\xc1\x8d\xba\x1e\xb6To\x1a\x07\xa4\xcaQ\x11qd\x19\xaf\xe7\x17B\x18F\xd3\xbfK\x1a03n\xb1\n\xa6&I\x1a\xef\x06`d\xa9-\x02\x9c\xc3?Sa\xbcu\xc1\x9aB\x948|\x8b\xd1\xbf"~\x04\xc2If\xfe\x81^\xb2$t\x16\np\x95\x9b\xb9\xfb\xeaX#v\x97\xd7\x02\xe4e\xa3\xdc\xb1o\x07\xcc\x93\xb3\xed\xda\xadh\xc2\xb5Z\x1cAk/\x19\xa1A\xed\n\tP\x01\xb7\xf5)\x03\xf2\xadX\x85\x83\x1d\xca\xd0\xd6I\'\x9a\xf6\x8e1o\x141f\xb0Y\x03^\xd9\xfd0\xc0G\x9a\x96\x92\x12\x02m\xdd\x1d\xf7OV^\xbc\x15b\xf3\xe9\x96\x86\x19O\xf5x\xafl<!=\xaa\x0fHD\xaeSc\x03\x1e\x02\xaa\x99Y\xb5E\x9d\x83\xcf\x00\xcdT\x0b\xb5\x18\xb4\x8bS\x9a\xac\xfe\xb3%\x8d\xd3\xb6h\xac_\x84\xe4\r\xcd\x03\xc9\xf6\x1c&68I)\xa3\x1a\x98\xd8C\xc3n\xa2\xc9e\xda\xae\x81e\xd3\xa0\x13IvT\xecf\xef\'\x03\xba\x0e\xa0\xa6e\x0eSR/\xad\xf4JHi\x11\xe4}\xe7\xfaZ\xa4\xdf\x04\x94\xbf\x02.\xb4\xcf\xc3\x92\xac'
ERROR:Controller:uncaught task exception: ushort format requires 0 <= number <= (0x7fff * 2 + 1)
Traceback (most recent call last):
  File "/home/electrum/electrumx/electrumx/server/controller.py", line 202, in check_task_exception
    task.result()
  File "/home/electrum/electrumx/electrumx/server/block_processor.py", line 203, in main_loop
    await task()
  File "/home/electrum/electrumx/electrumx/server/block_processor.py", line 247, in check_and_advance_blocks
    await self.controller.run_in_executor(self.advance_blocks, blocks)
  File "/home/electrum/electrumx/electrumx/server/controller.py", line 186, in run_in_executor
    return await self.loop.run_in_executor(None, func, *args)
  File "/usr/lib/python3.6/concurrent/futures/thread.py", line 56, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/electrum/electrumx/electrumx/server/block_processor.py", line 497, in advance_blocks
    undo_info = self.advance_txs(block.transactions)
  File "/home/electrum/electrumx/electrumx/server/block_processor.py", line 542, in advance_txs
    cache_value = spend_utxo(txin.prev_hash, txin.prev_idx)
  File "/home/electrum/electrumx/electrumx/server/block_processor.py", line 696, in spend_utxo
    idx_packed = pack('<H', tx_idx)
struct.error: ushort format requires 0 <= number <= (0x7fff * 2 + 1)
```

This happens for the reorg case as well.

Reply:
So your daemon is permitting invalid transactions?  Or these are valid and should be present but the inputs should be ignored?

Reply:
These inputs are actually valid, but represent a stealth input. Please see:

https://en.bitcoin.it/wiki/OP_RETURN

under "OP_RETURN applications".